test/test_rake_file_utils.rb

```diff
@@ -287,6 +287,110 @@ def test_sh_show_command
     assert_equal expected_cmd, show_cmd
   end
 
+  def test_sh_if_a_command_exits_with_error_status_it_raises_a_runtime_error_and_echoes_it_fully_long
+    assert_sh_echoes_it_fully(*string_and_output_and_error_long)
+  end
+
+  def test_sh_if_a_command_exits_with_error_status_it_raises_a_runtime_error_and_echoes_it_fully_short
+    assert_sh_echoes_it_fully(*string_and_output_and_error_short)
+  end
+
+  def test_sh_if_a_command_exits_with_error_status_it_raises_a_runtime_error_and_prints_its_full_output_when_verbose_long
+    verbose true
+    assert_sh_prints_its_full_output(*string_and_output_and_error_long)
+  end
+
+  def test_sh_if_a_command_exits_with_error_status_it_raises_a_runtime_error_and_prints_its_full_output_when_verbose_short
+    verbose true
+    assert_sh_prints_its_full_output(*string_and_output_and_error_short)
+  end
+
+  def test_sh_if_a_command_exits_with_error_status_then_providing_a_block_fully_echoes_a_command_when_verbose_long
+    string, output, error_message = string_and_output_and_error_long
+    command = build_command string, output, error_message
+    verbose true
+    actual = capture_subprocess_io do
+      sh(command){|ok, status|}
+    end
+    assert_output_equal actual, command, output, error_message
+  end
+
+  def assert_output_equal(actual, command, output, error_message)
+    # Unix standard-output, standard-error:
+    expected = [output, "#{command}\n#{error_message}"]
+    exp = standard_error_sort expected
+    act = standard_error_sort actual
+    assert_equal exp, act
+  end
+  private :assert_output_equal
+
+  def assert_sh_echoes_it_fully(string, output, error_message)
+    command = build_command string, output, error_message
+    # For testing purposes, we needn't match other programs' various
+    # error exit status code numbers, just Ruby's:
+    expected_error_message = "Command failed with status (#{error_exit_status}): [#{command}]"
+    ex = nil # Predefine to survive the block.
+    capture_subprocess_io do
+      ex = assert_raises RuntimeError do
+        sh command
+      end
+    end
+    assert_equal expected_error_message, ex.message
+  end
+  private :assert_sh_echoes_it_fully
+
+  def assert_sh_prints_its_full_output(string, output, error_message)
+    command = build_command string, output, error_message
+    actual = capture_subprocess_io do
+      assert_raises RuntimeError do
+        sh command
+      end
+    end
+    assert_output_equal actual, command, output, error_message
+  end
+  private :assert_sh_prints_its_full_output
+
+  def build_command(string, output, error_message)
+    clause_string = string.empty? ? '' : "v='#{string}';"
+    clause_stdout = output.empty? ? '' : "puts '#{output}';"
+    clause_stderr = error_message.empty? ? '' : "STDERR.puts '#{error_message}';"
```

Comment:
`''` and `nil` have the give the same result in interpolation so can you write these as:

`clause_string = "v='#{string}';" unless string.empty?`

It will be less confusing for maintainers.